### PR TITLE
Handle browser's Back button on scan page

### PR DESCRIPTION
**Pull request by KraPete**
## Release Notes

- **New Features**
	- Introduced a new scanning route (`/scan`) in the application for enhanced navigation.
	- Added mobile mode detection to improve responsiveness in the Device Manager page.
	- Enhanced device selection handling with new methods for better management.

- **Bug Fixes**
	- Improved functionality of the Back button on the scan page to enhance user navigation.

- **Refactor**
	- Streamlined methods for handling device selection and state transitions across various components.
	- Simplified control flow in the `NewDevicesScanPageStore` and `SearchDisconnectedScanPageStore`.

- **Documentation**
	- Updated type definitions and comments for better clarity and understanding of the codebase.

### CODE REVIEW

File: app/scripts/react-directives/device-manager/deviceManagerPage.jsx [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
 import NewDevicesScanPage from './scan/newDevicesScanPage.jsx';
 import ConfigEditorPage from './config-editor/configEditorPage.jsx';
 import SearchDisconnectedScanPage from './scan/searchDisconnectedScanPage.jsx';
+import { useMediaQuery } from 'react-responsive';
 
 const DeviceManagerPage = observer(({ pageStore }) => {
+  const checkMobile = useMediaQuery({ maxWidth: 991 });
+  useEffect(() => {
+    pageStore.setMobileMode(checkMobile);
+  });
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPage.jsx:12]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPage.jsx [hunk 2]
```diff
@@ -1,10 +1,15 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import { observer } from 'mobx-react-lite';
 import NewDevicesScanPage from './scan/newDevicesScanPage.jsx';
 import ConfigEditorPage from './config-editor/configEditorPage.jsx';
 import SearchDisconnectedScanPage from './scan/searchDisconnectedScanPage.jsx';
+import { useMediaQuery } from 'react-responsive';
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPage.jsx:6]: Useless

File: app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js [hunk 3]
```diff
@@ -50,21 +76,16 @@ class SearchDisconnectedScanPageStore {
...
 
-      this.onCancel = () => {
-        this.stopScanning();
-        resolve(undefined);
-      };
-    });
+  onCancel() {
+    this.stopScanning();
+    this?.onLeave(undefined);
   }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js:89]: Useless

File: app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js [hunk 4]
```diff
@@ -1,16 +1,42 @@
...
+  constructor(deviceManagerProxy, deviceTypesStore, onLeave) {
     this.commonScanStore = new CommonScanStore(deviceManagerProxy, deviceTypesStore);
-    this.onCancel = undefined;
-    this.onOk = undefined;
     this.active = false;
     this.portPath = undefined;
     this.deviceTypesStore = deviceTypesStore;
+    this.onLeave = onLeave;
 
     makeObservable(this, { active: observable, select: action, stopScanning: action });
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js:41]: Useless

File: app/scripts/react-directives/device-manager/scan/newDevicesScanPageStore.js [hunk 5]
```diff
@@ -41,40 +68,24 @@ class NewDevicesScanPageStore {
...
-    });
+  }
+
+  async onOk() {
+    try {
+      const devices = await this.confirmAddressChange();
+      this.stopScanning();
+      this?.onLeave(devices);
+    } catch (e) {}
+  }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/scan/newDevicesScanPageStore.js:84]: Useless

File: app/scripts/react-directives/device-manager/device-manager.js [hunk 6]
```diff
@@ -96,26 +117,49 @@ function deviceManagerDirective(
...
           scope.store.stopScanning();
         }
+
+        if (to == 'serial-config.properties') {
+          if (!scope.store.inMobileMode) {
+            return $state.target('serial-config');
+          }
+          scope.store.movedToDeviceProperties();
+        }
+
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/device-manager.js:162]: Useless

File: app/scripts/react-directives/device-manager/device-manager.js [hunk 7]
```diff
@@ -96,26 +117,49 @@ function deviceManagerDirective(
...
+        // Button hint is used to distinguish between direct URL and button click
         if (
-          transition.to().name == 'serial-config' &&
-          transition.from().name == 'serial-config.properties'
+          from == 'serial-config' &&
+          ['serial-config.properties', 'serial-config.scan'].includes(to) &&
+          !transition.params('to').hint
         ) {
+          return $state.target('serial-config');
+        }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/device-manager.js:132]: Useless

File: app/scripts/react-directives/device-manager/config-editor/tabsStore.js [hunk 8]
```diff
@@ -113,10 +113,14 @@ export class MobileModeTabsStore {
...
+  movedToContentPanel() {
+    this.tabsPanelIsActive = false;
+  }
+
   setMobileMode(value) {
     if (this.inMobileMode != value) {
       this.inMobileMode = value;
-      this.tabsPanelIsActive = true;
+      this.showTabsPanel();
     }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/config-editor/tabsStore.js:124]: Useless

File: app/scripts/app.routes.js [hunk 9]
```diff
@@ -316,6 +316,13 @@ function routing($stateProvider, $locationProvider, $urlRouterProvider) {
       },
     })
     //...........................................................................
+    .state('serial-config.scan', {
+      url: '/scan',
+      params: {
+        hint: false,
+      },
+    })
+    //...........................................................................
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/app.routes.js:325]: Useless

File: debian/changelog [hunk 10]
```diff
@@ -1,3 +1,9 @@
+wb-mqtt-homeui (2.106.0) stable; urgency=medium
+
+  * Handle browser's Back button on scan page in wb-mqtt-serial config editor
```
[the remark this answers is not in the document]
Comment (KraPete) [reply debian/changelog:3]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPageStore.js [hunk 11]
```diff
@@ -23,21 +22,25 @@ class DeviceManagerPageStore {
...
       this.deviceTypesStore,
       rolesFactory,
       setupDeviceFn,
       fwUpdateProxy
     );
     this.newDevicesScanPageStore = new NewDevicesScanPageStore(
       deviceManagerProxy,
-      this.deviceTypesStore
+      this.deviceTypesStore,
+      stateTransitions.onLeaveScan
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPageStore.js:35]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPageStore.js [hunk 12]
```diff
@@ -23,21 +22,25 @@ class DeviceManagerPageStore {
...
     );
     this.searchDisconnectedScanPageStore = new SearchDisconnectedScanPageStore(
       deviceManagerProxy,
-      this.deviceTypesStore
+      this.deviceTypesStore,
+      stateTransitions.onLeaveSearchDisconnectedDevice
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPageStore.js:40]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPageStore.js [hunk 13]
```diff
@@ -75,24 +82,33 @@ class DeviceManagerPageStore {
...
 
-  async addWbDevice() {
-    await this.configEditorPageStore.addDevices(
-      await this.newDevicesScanPageStore.select(
-        new ConfiguredDevices(this.configEditorPageStore.tabs.portTabs, this.deviceTypesStore)
-      )
+  addWbDevice() {
+    this.stateTransitions.toScan();
+    this.newDevicesScanPageStore.select(
+      new ConfiguredDevices(this.configEditorPageStore.tabs.portTabs, this.deviceTypesStore)
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPageStore.js:88]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPageStore.js [hunk 14]
```diff
@@ -75,24 +82,33 @@ class DeviceManagerPageStore {
...
+
+  searchDisconnectedDevice() {
+    const selectedDeviceTab = this.configEditorPageStore.tabs.selectedTab;
+    const selectedPortTab = this.configEditorPageStore.tabs.selectedPortTab;
+    this.stateTransitions.toScan();
+    this.searchDisconnectedScanPageStore.select(
+      selectedDeviceTab.deviceType,
+      selectedPortTab.editedData.path,
+      new ConfiguredDevices(this.configEditorPageStore.tabs.portTabs, this.deviceTypesStore),
+      selectedDeviceTab.slaveId
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPageStore.js:104]: Useless

File: app/scripts/react-directives/device-manager/deviceManagerPageStore.js [hunk 15]
```diff
@@ -75,24 +82,33 @@ class DeviceManagerPageStore {
...
   }
 
-  async searchDisconnectedDevice() {
+  addScannedDevices(selectedDevices) {
+    this.configEditorPageStore.addDevices(selectedDevices);
+  }
+
+  searchDisconnectedDevice() {
+    const selectedDeviceTab = this.configEditorPageStore.tabs.selectedTab;
+    const selectedPortTab = this.configEditorPageStore.tabs.selectedPortTab;
```
[the remark this answers is not in the document]
Comment (KraPete) [reply app/scripts/react-directives/device-manager/deviceManagerPageStore.js:98]: Useless

File: app/scripts/react-directives/device-manager/config-editor/configEditorPage.jsx [hunk 16]
```diff
@@ -263,10 +262,6 @@ const PageHeader = observer(
 );
 
 const ConfigEditorPage = observer(({ pageStore, onAddWbDevice, onSearchDisconnectedDevice }) => {
-  const checkMobile = useMediaQuery({ maxWidth: 991 });
```
Comment (ninedev-i) [inline app/scripts/react-directives/device-manager/config-editor/configEditorPage.jsx:266]: Ты это убрал из-за переезда логики в стор?

File: app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js [hunk 17]
```diff
@@ -50,21 +76,16 @@ class SearchDisconnectedScanPageStore {
...
-          runInAction(() => {
-            this.active = false;
-          });
-          resolve(this.commonScanStore.getSelectedDevices()[0]);
-        } catch (e) {}
-      };
+  }
+
+  onOk() {
+    this.stopScanning();
```
Comment (ninedev-i) [inline app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js:82]: не понял почему `onOk` останавливает сканирование

Comment (KraPete) [reply app/scripts/react-directives/device-manager/scan/searchDisconnectedScanPageStore.js]: Да, убрал

**Pull request merged**